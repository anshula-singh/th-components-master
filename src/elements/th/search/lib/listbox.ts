import { produce } from "immer";
import { ShadowRootTheGoodPart } from "lwc";

type Input = {
  onChange: (state: ListboxState) => void;
  onSelect: (index: number) => void;
  onSubmit: () => void;
  shadowRoot: ShadowRootTheGoodPart;
};

export type ListboxState = {
  active: boolean;
  selectedIndex: number;
  setup: boolean;
};

export type Selector = "root" | "input" | "options" | "option";

const SELECTOR_ROOT = "root";
const SELECTOR_INPUT = "input";
const SELECTOR_OPTIONS = "options";
const SELECTOR_OPTION = "option";

export function getSelector(selector: Selector) {
  return `[data-listbox="${selector}"]`;
}

export class Listbox {
  private _count: number = 0;
  private _state: ListboxState = {
    active: false,
    selectedIndex: -1,
    setup: false
  };
  private get state(): ListboxState {
    return this._state;
  }
  private set state(value: ListboxState) {
    this._state = value;
    this.update();
    this.input.onChange(this._state);
  }

  private hover = false;

  constructor(private input: Input) {
    input.onChange(this.state);
  }

  private setState(
    nextState: Partial<ListboxState> | ((state: ListboxState) => void)
  ) {
    this.state = produce(this.state, state => {
      if (typeof nextState === "function") nextState(state);
      else Object.assign(state, nextState);
    });
  }

  renderedCallback() {
    if (!this.state.setup) {
      let $input = this.getElement("input")!;
      let $options = this.getElement(SELECTOR_OPTIONS)!;
      $input.addEventListener("blur", this.onBlur);
      $input.addEventListener("focus", this.onFocus);
      $input.addEventListener("keydown", this.onKeyDown);
      $options.addEventListener("click", this.onClick);
      $options.addEventListener("mouseenter", this.onMouseEnter);
      $options.addEventListener("mouseleave", this.onMouseLeave);
      this.state.setup = true;
    }
    this._count = this.input.shadowRoot.querySelectorAll(
      getSelector(SELECTOR_OPTION)
    ).length;
    this.update();
  }

  disconnectedCallback() {
    let $input = this.getElement(SELECTOR_INPUT)!;
    let $options = this.getElement(SELECTOR_OPTIONS)!;
    $input.removeEventListener("blur", this.onBlur);
    $input.removeEventListener("focus", this.onFocus);
    $input.removeEventListener("keydown", this.onKeyDown);
    $options.removeEventListener("click", this.onClick);
    $options.removeEventListener("mouseenter", this.onMouseEnter);
    $options.removeEventListener("mouseleave", this.onMouseLeave);
  }

  private getElement(selector: Selector) {
    return this.input.shadowRoot.querySelector(getSelector(selector));
  }

  private onClick = (e: MouseEvent) => {
    let selectedIndex = Array.from(
      this.input.shadowRoot.querySelectorAll(getSelector(SELECTOR_OPTION))
    ).findIndex(n => n.contains(e.target as HTMLElement));
    this.setState({ active: false, selectedIndex });
    this.input.onSelect(this.state.selectedIndex);
  };

  private onBlur = (e: FocusEvent) => {
    if (!this.hover) {
      this.setState({ active: false });
    }
  };

  private onFocus = (e: FocusEvent) => {
    this.setState({ active: true, selectedIndex: -1 });
  };

  private onKeyDown = (e: KeyboardEvent) => {
    switch (e.key) {
      case "ArrowUp":
        this.setState(state => {
          state.selectedIndex--;
          if (state.selectedIndex < 0) state.selectedIndex = this._count - 1;
        });
        break;
      case "ArrowDown":
        this.setState(state => {
          state.selectedIndex++;
          if (state.selectedIndex === this._count) state.selectedIndex = 0;
        });
        break;
      case "Enter":
        this.setState({ active: false });
        if (this.state.selectedIndex > -1) {
          this.input.onSelect(this.state.selectedIndex);
        } else {
          this.input.onSubmit();
        }
        break;
    }
  };

  private onMouseEnter = (e: MouseEvent) => {
    this.hover = true;
  };

  private onMouseLeave = (e: MouseEvent) => {
    this.hover = false;
  };

  private setRootAttributes() {
    let $root = this.getElement(SELECTOR_ROOT)!;
    $root.setAttribute("aria-expanded", String(this.state.active));
    $root.setAttribute("aria-haspopup", "listbox");
    $root.setAttribute("role", "combobox");
  }

  private setInputAttributes() {
    let $activeDecendant = this.input.shadowRoot.querySelectorAll(
      getSelector(SELECTOR_OPTION)
    )[this.state.selectedIndex] as HTMLElement | undefined;
    let $input = this.getElement(SELECTOR_INPUT)!;
    let $options = this.getElement(SELECTOR_OPTIONS)!;
    $input.setAttribute(
      "aria-activedescendant",
      $activeDecendant ? $activeDecendant.getAttribute("id") || "" : ""
    );
    $input.setAttribute("aria-autocomplete", "list");
    $input.setAttribute("aria-controls", $options.getAttribute("id") || "");
    $input.setAttribute("autocomplete", "off");
  }

  private setListboxAttributes() {
    let $options = this.getElement(SELECTOR_OPTIONS)!;
    $options.setAttribute("role", "listbox");
    $options.setAttribute("aria-hidden", String(!this.state.active));
    this.input.shadowRoot.querySelectorAll("ui,li").forEach(node => {
      let element = node as HTMLElement;
      if (element.hasAttribute("role")) return;
      element.setAttribute("role", "presentation");
    });
  }

  private setListboxItemAttributes() {
    this.input.shadowRoot
      .querySelectorAll(getSelector(SELECTOR_OPTION))
      .forEach((node, i) => {
        let element = node as HTMLElement;
        element.setAttribute(
          "aria-selected",
          String(i === this.state.selectedIndex)
        );
        element.setAttribute("role", "option");
      });
  }

  private update() {
    this.setRootAttributes();
    this.setInputAttributes();
    this.setListboxAttributes();
    this.setListboxItemAttributes();
  }
}
