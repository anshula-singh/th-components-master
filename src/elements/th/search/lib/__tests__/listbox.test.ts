import { Listbox, ListboxState } from "../listbox";

jest.useFakeTimers();

describe("Listbox", () => {
  let $shadow: ShadowRoot;
  let $input: HTMLInputElement;
  let listbox: Listbox;
  let onChange: jest.Mock;
  let onSelect: jest.Mock;
  let onSubmit: jest.Mock;
  afterEach(() => {
    while (document.body.firstChild) {
      document.body.removeChild(document.body.firstChild);
    }
  });
  beforeEach(() => {
    onChange = jest.fn();
    onSelect = jest.fn();
    onSubmit = jest.fn();
    let $element = document.createElement("div");
    $shadow = $element.attachShadow({ mode: "open" });
    $shadow.innerHTML = createListboxTemplate(createListboxItems());
    listbox = createListbox($shadow, onChange, onSelect, onSubmit);
    listbox.renderedCallback();
    $input = $shadow.querySelector("input") as HTMLInputElement;
  });
  it("default", async () => {
    expect($shadow.innerHTML).toMatchSnapshot();
  });
  it("focus", async () => {
    let focusEvent = new FocusEvent("focus", {
      bubbles: false,
      cancelable: true
    });
    $input.dispatchEvent(focusEvent);
    expect($shadow.innerHTML).toMatchSnapshot();
  });
  it("submit", async () => {
    let focusEvent = new FocusEvent("focus", {
      bubbles: false,
      cancelable: true
    });
    let keydownEvent = new KeyboardEvent("keydown", {
      bubbles: false,
      cancelable: true,
      key: "Enter"
    });
    $input.dispatchEvent(focusEvent);
    $input.value = "hello";
    $input.dispatchEvent(keydownEvent);
    expect(onSubmit).toHaveBeenCalledTimes(1);
  });
  it("arrow down", async () => {
    let focusEvent = new FocusEvent("focus", {
      bubbles: false,
      cancelable: true
    });
    let keydownEvent = new KeyboardEvent("keydown", {
      bubbles: false,
      cancelable: true,
      key: "ArrowDown"
    });
    $input.dispatchEvent(focusEvent);
    $input.dispatchEvent(keydownEvent);
    expect($input).toHaveAttribute("aria-activedescendant", "a");
    $input.dispatchEvent(keydownEvent);
    expect($input).toHaveAttribute("aria-activedescendant", "b");
    $input.dispatchEvent(keydownEvent);
    expect($input).toHaveAttribute("aria-activedescendant", "c");
    $input.dispatchEvent(keydownEvent);
    expect($input).toHaveAttribute("aria-activedescendant", "a");
  });
  it("arrow up", async () => {
    let focusEvent = new FocusEvent("focus", {
      bubbles: false,
      cancelable: true
    });
    let keydownEvent = new KeyboardEvent("keydown", {
      bubbles: false,
      cancelable: true,
      key: "ArrowUp"
    });
    $input.dispatchEvent(focusEvent);
    $input.dispatchEvent(keydownEvent);
    expect($input).toHaveAttribute("aria-activedescendant", "c");
    $input.dispatchEvent(keydownEvent);
    expect($input).toHaveAttribute("aria-activedescendant", "b");
  });
  it("enter", async () => {
    let focusEvent = new FocusEvent("focus", {
      bubbles: false,
      cancelable: true
    });
    let keydownEvent = (key: string) =>
      new KeyboardEvent("keydown", {
        bubbles: false,
        cancelable: true,
        key
      });
    $input.dispatchEvent(focusEvent);
    $input.dispatchEvent(keydownEvent("ArrowDown"));
    $input.dispatchEvent(keydownEvent("ArrowDown"));
    $input.dispatchEvent(keydownEvent("Enter"));
    expect(onSelect).toHaveBeenCalledWith(1);
  });
  it("blur", async () => {
    let focusEvent = new FocusEvent("blur", {
      bubbles: false,
      cancelable: true
    });
    $input.dispatchEvent(focusEvent);
    expect($shadow.innerHTML).toMatchSnapshot();
  });
  it("click", async () => {
    let focusEvent = new FocusEvent("blur", {
      bubbles: false,
      cancelable: true
    });
    let mouseEnterEvent = new MouseEvent("mouseenter", {
      bubbles: false,
      cancelable: true
    });
    let clickEvent = new MouseEvent("click", {
      bubbles: true,
      cancelable: true
    });
    $input.dispatchEvent(focusEvent);
    $shadow
      .querySelector('[data-listbox="options"]')!
      .dispatchEvent(mouseEnterEvent);
    $shadow
      .querySelectorAll('[data-listbox="option"]')[1]
      .dispatchEvent(clickEvent);
    expect(onSelect).toHaveBeenCalledWith(1);
  });
});

const createListboxTemplate = (items: { key: string; value: string }[]) => `
  <div data-listbox="root">
    <div>
      <input data-listbox="input" />
    </div>
    <div data-listbox="options">
      <ul>
        ${items.map(
          item => `
          <li>
            <div data-listbox="option" id="${item.key}">
              ${item.value}
            </div>
          </li>
        `
        )}
      </ul>
    </div>
  </div>
`;

const createListboxItems = () => [
  { key: "a", value: "A" },
  { key: "b", value: "B" },
  { key: "c", value: "C" }
];

const createListbox = (
  shadowRoot: ShadowRoot,
  onChange: (state: ListboxState) => void,
  onSelect: (index: number) => void,
  onSubmit: () => void
) =>
  new Listbox({
    onChange,
    onSelect,
    onSubmit,
    // @ts-ignore
    shadowRoot: shadowRoot
  });
