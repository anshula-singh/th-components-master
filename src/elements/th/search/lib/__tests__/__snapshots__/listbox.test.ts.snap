// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Listbox blur 1`] = `
"
  <div data-listbox=\\"root\\" aria-expanded=\\"false\\" aria-haspopup=\\"listbox\\" role=\\"combobox\\">
    <div>
      <input data-listbox=\\"input\\" aria-activedescendant=\\"\\" aria-autocomplete=\\"list\\" aria-controls=\\"\\" autocomplete=\\"off\\">
    </div>
    <div data-listbox=\\"options\\" role=\\"listbox\\" aria-hidden=\\"true\\">
      <ul>
        
          <li role=\\"presentation\\">
            <div data-listbox=\\"option\\" id=\\"a\\" aria-selected=\\"false\\" role=\\"option\\">
              A
            </div>
          </li>
        ,
          <li role=\\"presentation\\">
            <div data-listbox=\\"option\\" id=\\"b\\" aria-selected=\\"false\\" role=\\"option\\">
              B
            </div>
          </li>
        ,
          <li role=\\"presentation\\">
            <div data-listbox=\\"option\\" id=\\"c\\" aria-selected=\\"false\\" role=\\"option\\">
              C
            </div>
          </li>
        
      </ul>
    </div>
  </div>
"
`;

exports[`Listbox default 1`] = `
"
  <div data-listbox=\\"root\\" aria-expanded=\\"false\\" aria-haspopup=\\"listbox\\" role=\\"combobox\\">
    <div>
      <input data-listbox=\\"input\\" aria-activedescendant=\\"\\" aria-autocomplete=\\"list\\" aria-controls=\\"\\" autocomplete=\\"off\\">
    </div>
    <div data-listbox=\\"options\\" role=\\"listbox\\" aria-hidden=\\"true\\">
      <ul>
        
          <li role=\\"presentation\\">
            <div data-listbox=\\"option\\" id=\\"a\\" aria-selected=\\"false\\" role=\\"option\\">
              A
            </div>
          </li>
        ,
          <li role=\\"presentation\\">
            <div data-listbox=\\"option\\" id=\\"b\\" aria-selected=\\"false\\" role=\\"option\\">
              B
            </div>
          </li>
        ,
          <li role=\\"presentation\\">
            <div data-listbox=\\"option\\" id=\\"c\\" aria-selected=\\"false\\" role=\\"option\\">
              C
            </div>
          </li>
        
      </ul>
    </div>
  </div>
"
`;

exports[`Listbox focus 1`] = `
"
  <div data-listbox=\\"root\\" aria-expanded=\\"true\\" aria-haspopup=\\"listbox\\" role=\\"combobox\\">
    <div>
      <input data-listbox=\\"input\\" aria-activedescendant=\\"\\" aria-autocomplete=\\"list\\" aria-controls=\\"\\" autocomplete=\\"off\\">
    </div>
    <div data-listbox=\\"options\\" role=\\"listbox\\" aria-hidden=\\"false\\">
      <ul>
        
          <li role=\\"presentation\\">
            <div data-listbox=\\"option\\" id=\\"a\\" aria-selected=\\"false\\" role=\\"option\\">
              A
            </div>
          </li>
        ,
          <li role=\\"presentation\\">
            <div data-listbox=\\"option\\" id=\\"b\\" aria-selected=\\"false\\" role=\\"option\\">
              B
            </div>
          </li>
        ,
          <li role=\\"presentation\\">
            <div data-listbox=\\"option\\" id=\\"c\\" aria-selected=\\"false\\" role=\\"option\\">
              C
            </div>
          </li>
        
      </ul>
    </div>
  </div>
"
`;
